File: api_explorer/reporter.py
from __future__ import annotations

import json
from typing import Any
from urllib.parse import urlparse

from .models import ApiCallResult, EndpointTestResult, TestSuiteResult

R = "\033[0m"
B = "\033[1m"
D = "\033[2m"
RED = "\033[31m"
GRN = "\033[32m"
YLW = "\033[33m"
BLU = "\033[34m"
CYN = "\033[36m"

H_LINE = "\u2501" * 60
T_LINE = "\u2500" * 56
PIPE = "\u2502"
CORNER_TL = "\u250c"
CORNER_BL = "\u2514"
DASH = "\u2500"
ARROW = "\u21b3"
MDASH = "\u2014"
SKIP = "\u2298"
CHECK = "\u2713"
CROSS = "\u2717"
SEP_86 = "  " + DASH * 86
SEP_52 = "  " + DASH * 52
FOOTER_LINE = DASH * 59


def print_header() -> None:
    print(f"\n{B}{CYN}{H_LINE}{R}")
    print(f"{B}{CYN}  🏠 Samsung SmartThings API 테스트{R}")
    print(f"{B}{CYN}{H_LINE}{R}\n")


def print_category_header(label: str) -> None:
    pad = DASH * max(0, 50 - len(label))
    print(f"\n{B}{BLU}{CORNER_TL}{DASH} {label} {pad}{R}")


def print_category_footer() -> None:
    print(f"{B}{PIPE}{R}")
    print(f"{B}{CORNER_BL}{FOOTER_LINE}{R}")


def print_test_result(result: EndpointTestResult) -> None:
    if result.skipped:
        icon = f"{YLW}{SKIP}{R}"
    elif result.success:
        icon = f"{GRN}{CHECK}{R}"
    else:
        icon = f"{RED}{CROSS}{R}"

    print(f"{B}{PIPE}{R}")
    print(
        f"{B}{PIPE}{R} {icon} {B}{result.endpoint.name}{R} {D}{MDASH} {result.endpoint.description}{R}"
    )

    if result.skipped:
        reason = result.endpoint.needs_setup or "부수효과 테스트 건너뜀"
        print(f"{B}{PIPE}{R}   {YLW}건너뜀: {reason}{R}")
        return

    if result.error:
        print(f"{B}{PIPE}{R}   {RED}오류: {result.error}{R}")

    for call in result.calls:
        _print_api_call(call)


def _print_api_call(call: ApiCallResult) -> None:
    status = call.response["status"]
    sc = GRN if status < 300 else (YLW if status < 400 else RED)

    print(f"{B}{PIPE}{R}   {D}{T_LINE}{R}")
    print(
        f"{B}{PIPE}{R}   {CYN}요청:{R} {B}{call.request['method']}{R} {call.request['url']}"
    )

    if call.request.get("body"):
        print(f"{B}{PIPE}{R}   {CYN}요청 본문:{R}")
        _fmt_json(call.request["body"], 8)

    print(
        f"{B}{PIPE}{R}   {CYN}응답:{R} {sc}{status} {call.response['status_text']}{R} {D}({call.duration}ms){R}"
    )

    body = call.response.get("body")
    if body is not None:
        print(f"{B}{PIPE}{R}   {CYN}응답 데이터:{R}")
        _fmt_json(body, 30)


def _fmt_json(data: Any, max_lines: int) -> None:
    lines = json.dumps(data, indent=2, ensure_ascii=False).split("\n")
    for line in lines[:max_lines]:
        print(f"{B}{PIPE}{R}     {line}")
    if len(lines) > max_lines:
        remaining = len(lines) - max_lines
        print(f"{B}{PIPE}{R}     {D}... ({remaining}줄 생략){R}")


def print_detail_table(suite: TestSuiteResult) -> None:
    print(f"\n{B}{CYN}{H_LINE}{R}")
    print(f"{B}{CYN}  📋 엔드포인트별 상세 결과{R}")
    print(f"{B}{CYN}{H_LINE}{R}\n")

    print(SEP_86)
    hdr = f"  {B}{'#':<4}{'결과':<7}{'카테고리':<16}{'테스트명':<24}{'메서드':<8}{'응답':<8}{'시간':<10}경로{R}"
    print(hdr)
    print(SEP_86)

    for i, r in enumerate(suite.results, 1):
        if r.skipped:
            icon, label = f"{YLW}{SKIP}{R}", "건너뜀"
        elif r.success:
            icon, label = f"{GRN}{CHECK}{R}", "성공"
        else:
            icon, label = f"{RED}{CROSS}{R}", "실패"

        fc = r.calls[0] if r.calls else None
        method = fc.request["method"] if fc else MDASH
        code = str(fc.response["status"]) if fc else MDASH
        dur = f"{fc.duration}ms" if fc else MDASH
        sc = (
            D
            if not fc
            else (
                GRN
                if fc.response["status"] < 300
                else (YLW if fc.response["status"] < 400 else RED)
            )
        )

        path = MDASH
        if fc:
            parsed = urlparse(fc.request["url"])
            q = f"?{parsed.query}" if parsed.query else ""
            path = parsed.path.replace("/v1", "") + q

        cat = r.endpoint.category
        name = r.endpoint.name

        print(
            f"  {D}{i:<4}{R}{icon} {label:<5} {cat:<16}{name:<24}{B}{method:<8}{R}{sc}{code:<8}{R}{D}{dur:<10}{R}{D}{path}{R}"
        )

        extra_label = f"{ARROW} 추가 호출"
        for cc in r.calls[1:]:
            csc = (
                GRN
                if cc.response["status"] < 300
                else (YLW if cc.response["status"] < 400 else RED)
            )
            cp = urlparse(cc.request["url"])
            cq = f"?{cp.query}" if cp.query else ""
            cpath = cp.path.replace("/v1", "") + cq
            cdur = f"{cc.duration}ms"
            print(
                f"  {'':4}{'':7}{'':16}{extra_label:<24}{B}{cc.request['method']:<8}{R}{csc}{cc.response['status']:<8}{R}{D}{cdur:<10}{R}{D}{cpath}{R}"
            )

        if r.error:
            print(f"  {'':4}{'':7}{RED}{ARROW} 오류: {r.error}{R}")

    print(SEP_86)


def print_summary(suite: TestSuiteResult) -> None:
    print(f"\n{B}{CYN}{H_LINE}{R}")
    print(f"{B}{CYN}  📊 테스트 결과 요약{R}")
    print(f"{B}{CYN}{H_LINE}{R}\n")

    dur_sec = suite.duration / 1000
    print(f"  총 소요 시간: {B}{dur_sec:.1f}초{R}")
    print(f"  전체: {B}{suite.total}{R}개")
    print(f"  {GRN}성공: {suite.passed}{R}개")
    print(f"  {RED}실패: {suite.failed}{R}개")
    if suite.skipped > 0:
        print(f"  {YLW}건너뜀: {suite.skipped}{R}개")

    cats: dict[str, dict[str, int]] = {}
    for r in suite.results:
        cat = r.endpoint.category_label
        s = cats.setdefault(cat, {"total": 0, "passed": 0, "failed": 0, "skipped": 0})
        s["total"] += 1
        if r.skipped:
            s["skipped"] += 1
        elif r.success:
            s["passed"] += 1
        else:
            s["failed"] += 1

    print(f"\n  {B}카테고리별 결과:{R}")
    print(SEP_52)
    h_cat = "카테고리"
    h_pass = "성공"
    h_fail = "실패"
    h_skip = "건너뜀"
    print(f"  {B}{h_cat:<24}{h_pass:<8}{h_fail:<8}{h_skip:<8}전체{R}")
    print(SEP_52)
    for cat, s in cats.items():
        pad = cat + " " * max(0, 22 - len(cat))
        print(
            f"  {pad}  {GRN}{s['passed']:<8}{R}{RED}{s['failed']:<8}{R}{YLW}{s['skipped']:<8}{R}{s['total']}"
        )
    print(SEP_52 + "\n")

File: api_explorer_rs/src/reporter.rs
use serde_json::Value;

use crate::models::{ApiCallResult, EndpointTestResult, TestSuiteResult};

const R: &str = "\x1b[0m";
const B: &str = "\x1b[1m";
const D: &str = "\x1b[2m";
const RED: &str = "\x1b[31m";
const GRN: &str = "\x1b[32m";
const YLW: &str = "\x1b[33m";
const BLU: &str = "\x1b[34m";
const CYN: &str = "\x1b[36m";

const H_LINE: &str = "━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━";
const T_LINE: &str = "────────────────────────────────────────────────────────";
const PIPE: &str = "│";
const CORNER_TL: &str = "┌";
const CORNER_BL: &str = "└";
const DASH: &str = "─";
const ARROW: &str = "↳";
const MDASH: &str = "—";
const SKIP: &str = "⊘";
const CHECK: &str = "✓";
const CROSS: &str = "✗";

pub fn print_header() {
    println!("\n{B}{CYN}{H_LINE}{R}");
    println!("{B}{CYN}  🏠 Samsung SmartThings API 테스트{R}");
    println!("{B}{CYN}{H_LINE}{R}\n");
}

pub fn print_category_header(label: &str) {
    let pad_len = 50usize.saturating_sub(label.chars().count());
    let pad = DASH.repeat(pad_len);
    println!("\n{B}{BLU}{CORNER_TL}{DASH} {label} {pad}{R}");
}

pub fn print_category_footer() {
    println!("{B}{PIPE}{R}");
    println!("{B}{CORNER_BL}{}{R}", DASH.repeat(59));
}

pub fn print_test_result(result: &EndpointTestResult) {
    let icon = if result.skipped {
        format!("{YLW}{SKIP}{R}")
    } else if result.success {
        format!("{GRN}{CHECK}{R}")
    } else {
        format!("{RED}{CROSS}{R}")
    };

    println!("{B}{PIPE}{R}");
    println!(
        "{B}{PIPE}{R} {icon} {B}{}{R} {D}{MDASH} {}{R}",
        result.endpoint.name, result.endpoint.description
    );

    if result.skipped {
        let reason = result
            .endpoint
            .needs_setup
            .clone()
            .unwrap_or_else(|| "부수효과 테스트 건너뜀".to_string());
        println!("{B}{PIPE}{R}   {YLW}건너뜀: {reason}{R}");
        return;
    }

    if let Some(error) = &result.error {
        println!("{B}{PIPE}{R}   {RED}오류: {error}{R}");
    }

    for call in &result.calls {
        print_api_call(call);
    }
}

fn print_api_call(call: &ApiCallResult) {
    let status = call.response.status;
    let sc = if status < 300 {
        GRN
    } else if status < 400 {
        YLW
    } else {
        RED
    };

    println!("{B}{PIPE}{R}   {D}{T_LINE}{R}");
    println!(
        "{B}{PIPE}{R}   {CYN}요청:{R} {B}{}{R} {}",
        call.request.method, call.request.url
    );

    if let Some(body) = &call.request.body {
        println!("{B}{PIPE}{R}   {CYN}요청 본문:{R}");
        fmt_json(body, 8);
    }

    println!(
        "{B}{PIPE}{R}   {CYN}응답:{R} {sc}{} {}{R} {D}({}ms){R}",
        call.response.status, call.response.status_text, call.duration_ms
    );

    if let Some(body) = &call.response.body {
        println!("{B}{PIPE}{R}   {CYN}응답 데이터:{R}");
        fmt_json(body, 30);
    }
}

fn fmt_json(data: &Value, max_lines: usize) {
    let rendered = serde_json::to_string_pretty(data).unwrap_or_else(|_| "null".to_string());
    let lines: Vec<&str> = rendered.lines().collect();

    for line in lines.iter().take(max_lines) {
        println!("{B}{PIPE}{R}     {line}");
    }

    if lines.len() > max_lines {
        let remaining = lines.len() - max_lines;
        println!("{B}{PIPE}{R}     {D}... ({remaining}줄 생략){R}");
    }
}

pub fn print_detail_table(suite: &TestSuiteResult) {
    println!("\n{B}{CYN}{H_LINE}{R}");
    println!("{B}{CYN}  📋 엔드포인트별 상세 결과{R}");
    println!("{B}{CYN}{H_LINE}{R}\n");

    let sep_86 = format!("  {}", DASH.repeat(86));
    println!("{sep_86}");
    println!(
        "  {B}{:<4}{:<7}{:<16}{:<24}{:<8}{:<8}{:<10}경로{R}",
        "#", "결과", "카테고리", "테스트명", "메서드", "응답", "시간"
    );
    println!("{sep_86}");

    for (i, r) in suite.results.iter().enumerate() {
        let (icon, label) = if r.skipped {
            (format!("{YLW}{SKIP}{R}"), "건너뜀")
        } else if r.success {
            (format!("{GRN}{CHECK}{R}"), "성공")
        } else {
            (format!("{RED}{CROSS}{R}"), "실패")
        };

        let first_call = r.calls.first();
        let method = first_call
            .map(|c| c.request.method.clone())
            .unwrap_or_else(|| MDASH.to_string());
        let code = first_call
            .map(|c| c.response.status.to_string())
            .unwrap_or_else(|| MDASH.to_string());
        let dur = first_call
            .map(|c| format!("{}ms", c.duration_ms))
            .unwrap_or_else(|| MDASH.to_string());

        let sc = if let Some(fc) = first_call {
            if fc.response.status < 300 {
                GRN
            } else if fc.response.status < 400 {
                YLW
            } else {
                RED
            }
        } else {
            D
        };

        let path = first_call
            .map(|c| extract_path(&c.request.url))
            .unwrap_or_else(|| MDASH.to_string());

        println!(
            "  {D}{:<4}{R}{} {:<5} {:<16}{:<24}{B}{:<8}{R}{}{:<8}{R}{D}{:<10}{R}{D}{}{R}",
            i + 1,
            icon,
            label,
            r.endpoint.category,
            r.endpoint.name,
            method,
            sc,
            code,
            dur,
            path
        );

        for call in r.calls.iter().skip(1) {
            let csc = if call.response.status < 300 {
                GRN
            } else if call.response.status < 400 {
                YLW
            } else {
                RED
            };

            println!(
                "  {:4}{:7}{:16}{:<24}{B}{:<8}{R}{}{:<8}{R}{D}{:<10}{R}{D}{}{R}",
                "",
                "",
                "",
                format!("{ARROW} 추가 호출"),
                call.request.method,
                csc,
                call.response.status,
                format!("{}ms", call.duration_ms),
                extract_path(&call.request.url)
            );
        }

        if let Some(err) = &r.error {
            println!("  {:4}{:7}{RED}{ARROW} 오류: {err}{R}", "", "");
        }
    }

    println!("{sep_86}");
}

pub fn print_summary(suite: &TestSuiteResult) {
    println!("\n{B}{CYN}{H_LINE}{R}");
    println!("{B}{CYN}  📊 테스트 결과 요약{R}");
    println!("{B}{CYN}{H_LINE}{R}\n");

    let duration_sec = suite.duration_ms as f64 / 1000.0;
    println!("  총 소요 시간: {B}{duration_sec:.1}초{R}");
    println!("  전체: {B}{}{R}개", suite.total);
    println!("  {GRN}성공: {}{R}개", suite.passed);
    println!("  {RED}실패: {}{R}개", suite.failed);
    if suite.skipped > 0 {
        println!("  {YLW}건너뜀: {}{R}개", suite.skipped);
    }

    let mut cats: Vec<(String, (usize, usize, usize, usize))> = Vec::new();
    for r in &suite.results {
        let label = r.endpoint.category_label.clone();
        if let Some((_, counts)) = cats.iter_mut().find(|(k, _)| *k == label) {
            counts.0 += 1;
            if r.skipped {
                counts.3 += 1;
            } else if r.success {
                counts.1 += 1;
            } else {
                counts.2 += 1;
            }
        } else {
            let mut counts = (1usize, 0usize, 0usize, 0usize);
            if r.skipped {
                counts.3 = 1;
            } else if r.success {
                counts.1 = 1;
            } else {
                counts.2 = 1;
            }
            cats.push((label, counts));
        }
    }

    println!("\n  {B}카테고리별 결과:{R}");
    let sep_52 = format!("  {}", DASH.repeat(52));
    println!("{sep_52}");
    println!(
        "  {B}{:<24}{:<8}{:<8}{:<8}전체{R}",
        "카테고리", "성공", "실패", "건너뜀"
    );
    println!("{sep_52}");
    for (cat, (total, passed, failed, skipped)) in cats {
        let pad = format!("{}{}", cat, " ".repeat(22usize.saturating_sub(cat.chars().count())));
        println!(
            "  {}  {GRN}{:<8}{R}{RED}{:<8}{R}{YLW}{:<8}{R}{}",
            pad, passed, failed, skipped, total
        );
    }
    println!("{}\n", sep_52);
}

fn extract_path(url: &str) -> String {
    let path_with_query = if let Some(scheme_pos) = url.find("://") {
        let rest = &url[(scheme_pos + 3)..];
        if let Some(path_idx) = rest.find('/') {
            rest[path_idx..].to_string()
        } else {
            "/".to_string()
        }
    } else {
        url.to_string()
    };

    if let Some(stripped) = path_with_query.strip_prefix("/v1") {
        stripped.to_string()
    } else {
        path_with_query
    }
}
